rust/headless-client/src/standalone.rs

```diff
@@ -170,47 +170,51 @@ pub fn run_only_headless_client() -> Result<()> {
     platform::notify_service_controller()?;
 
     let result = rt.block_on(async {
+        let mut terminate = signals::Terminate::new()?;
+        let mut hangup = signals::Hangup::new()?;
+        let mut terminate = pin!(terminate.recv().fuse());
+        let mut hangup = pin!(hangup.recv().fuse());
         let mut dns_controller = DnsController::default();
         let mut tun_device = tun_device_manager::TunDeviceManager::new()?;
-        let mut signals = Signals::new()?;
+        let mut cb_rx = ReceiverStream::new(cb_rx).fuse();
 
         loop {
-            match future::select(pin!(signals.recv()), pin!(cb_rx.recv())).await {
-                future::Either::Left((SignalKind::Hangup, _)) => {
+            let cb = futures::select! {
+                () = terminate => {
+                    tracing::info!("Caught SIGINT / SIGTERM / Ctrl+C");
+                    return Ok(());
+                },
+                () = hangup => {
                     tracing::info!("Caught SIGHUP");
                     session.reconnect();
+                    continue;
+                },
+                cb = cb_rx.next() => cb,
+            };
+
+            let Some(cb) = cb else {
+                bail!("cb_rx unexpectedly ran empty");
+            };
```

Comment:
```suggestion
                cb = cb_rx.next() => {
                    cb.context("cb_rx unexpectedly ran empty")?
                },
            };
```

Is that better? Not sure :man_shrugging: :)